cpu/arm7_common/Makefile.features

```diff
@@ -1,5 +1,5 @@
-CPU_ARCH = armv4t
-CPU_CORE = arm7tdmi_s
+CPU_ARCH := armv4t
+CPU_CORE := arm7tdmi_s
```

Comment:
Why do we need immediate expansion here?

Reply:
You should ask the other way round: Why do we need recursive expansion? We decided to use immediate expansion, unless there are good reasons to differently.

Reply:
> We decided to use immediate expansion

Who is we? https://doc.riot-os.org/build-system-basics.html#board-cpu-features, https://doc.riot-os.org/build-system-basics.html#board-cpu-features, https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421, https://github.com/RIOT-OS/RIOT/issues/8913, https://github.com/RIOT-OS/RIOT/issues/10850. AFAIK, past efforts were header the other way.

Reply:
The argument raised in https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421 only applies for non-const variables though.
It makes sense to use `:=` for constants and only for constants.

Reply:
Here it does not matter BTW, but seems we have contradicting impressions on the default behavior to adopt.

Reply:
> The argument raised in [#11099 (comment)](https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289820421) only applies to non-const variables.
> It makes sense to use `:=` for constants and only for constants.

Not necessarily, you might still want to override a constant, if you do `CPU_MODEL := foo`, them `CPU_MODULE=bar` won't override it. Again here it probably does not matter, no one is going to do `CPU_ARCH=far make`., and `CPU_MODEL=foo` has the same effect in this context, and both can be overridden by make arguments.

I also meant to reference this comment https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289842650.

Reply:
I'm not trying to bikeshed, the PR looks fine, just seems we are rowing in opposite directions, and I would rather we align :)

Reply:
> := has some disadvantages:
>
> * The value of variables defined with := depends on the order they are defined.

That clearly is a pro, not a con. A simple example:

```Makefile
a = foo
b = $(a)
a += bar
```

It is absolutely contrary to the behavior in every sane programming language, that `b` now has the value `foo bar`, and not `foo`.

> Here it does not matter BTW, but seems we have contradicting impressions on the default behavior to adopt.

We had this discussion a few times now. The first time was when the `make info-boards-supported` was taking minutes to completed with CPU usage at 100%, because of the explosive growth of complexity when expressions start to depend on each other. Switching to immediate expansion restored not only normal build time, but also forced some structuring that resulted in much clearer `Makefile` (or, less horrible, to be honest).

I was under the impression that each time we agreed on using immediate expansion by default.

Reply:
> I also meant to reference this comment [#11099 (comment)](https://github.com/RIOT-OS/RIOT/pull/11099#discussion_r289842650).

Honestly, recursive evaluation played a big role in why some parts of the build system are so hard to follow and horrible to debug: For understanding what a specific expression does, you do not only need to know what the contents of every referenced variable currently consist of, you also have to track what they *will* look like *every time* the variable is referenced in the future. With the complexity of RIOT's build system, this is impossible to follow. And this is why people start screaming when somebody touches the build system, as everybody knows this is likely to break things at places that seem totally unrelated for reasons nobody could have even dreamed of before.

Using immediate evaluations instead makes things predictable. And I think we all agree that a more predictable build system would greatly benefit us all.

And the argument regarding the better performance of recursive expansion: The exact opposite is true.

Reply:
> That clearly is a pro, not a con. A simple example:

It depends, on a recursive build system when the order of inclusion is not clear. Because you will have different behavior in different scenarios.

> much clearer Makefile (or, less horrible, to be honest).

I completely agree here, having it not be horrible would require someone constantly working on build-system stuff IMO.

>We had this discussion a few times now. The first time was when the make info-boards-supported was taking minutes to complete with CPU usage at 100%, because of the explosive growth of complexity when expressions start to depend on each other.

I'm in no case arguing against the use of `:=` in that situation. But the thing is `:=` can also have an overhead, and the question is what is the best default, not for someone that has suffered from the build system enough to decide and read make documentation, but for programmer/contributor lambda, that is focused on getting his c program to compile.

> I was under the impression that each time we agreed on using immediate expansion by default.

I think I'm still formatted to the default of not liking when I see a `:=`, also because when I started looking at build system stuff order of inclusion was still a huge issue, maybe that is no longer the bigger issue and I need to be re-formatted hahaha.

Anyway,  seems both of you think `:=` is the best default. So I'll shut up.